### PR TITLE
feat: lily chain export command

### LINKED ISSUE #697: Support chain exports
### Description

Add an export command to lily, like the one in Lotus but modernised. The Lotus version has some obscure options and has some custom async behaviour to avoid blocking. Lily could just run it as a specialised walk job.

The output would be a car file.

The export would start from a given height or the head of the chain and would include:
 - all block headers from starting height to genesis
 - optionally: all messages and receipts from  starting height to a given height
 - optionally: full state trees from  starting height to a given height

### PR DESCRIPTION
**Pull request by frrist**
- closes #697 
```
NAME:
   lily export - 

USAGE:
   
Exported chains will include all block headers starting at height 'from' to the genesis block.
This means block headers are not filtered by the 'from' and 'to' flags.
Messages, Receipts, and StateRoots are filtered by the 'from' and 'to' flags.

Some examples:

    lily export --from=100 --to=200 --include-messages=true --include-receipts=true --include-stateroots=false
        - export blocks from 200 to 0.
        - export messages from 200 to 100.
        - export receipts from 200 to 100.
        - no stateroots exported.

    lily export --repo=~/.lily --from=0 --to=200 --include-messages=true --include-receipts=true --include-stateroots=true:
        - export blocks from 200 to 0
        - export messages from 200 to 0
        - export receipts from 200 to 0
        - export stateroots from 200 to 0

    lily export --repo=~/.lily --from=0 --to=200 --include-messages=false --include-receipts=false --include-stateroots=false:
        - export all blocks from 200 to 0
        - no messages exported
        - no receipts exported
        - no stateroots exported


DESCRIPTION:
   Export chain from repo (requires node to be offline)

OPTIONS:
   --repo value          the repo to export chain from (default: "~/.lily")
   --to value            inclusive highest epoch to export (default: 0)
   --from value          inclusive lowest epoch to export (default: 0)
   --include-messages    exports messages if true (default: true)
   --include-receipts    exports receipts if true (default: true)
   --include-stateroots  exports stateroots if true (default: true)
   --out-file value      file to export to (default: "chain_export.car")
   --progress            set to show progress bar of export (default: true)
   --help, -h            show help (default: false)
   
```

```
lily export --repo=~/.lily --min=0 --max=10000 --include-messages=true --include-receipts=true --include-stateroots=true
```
```
INFO    lily/commands   commands/export.go:110  loading export head...
INFO    lily/commands   commands/export.go:116  loaded export head      {"tipset": "[bafy2bzacecnrejmgqegtqob5s2ut2vdqimcvmaxtjngpbrjbz2niztdcdj3vq bafy2bzaceazs7kvjm5uxknrheyl3qpwrol6ioi52c6ckamm7y7hawhkijjulk]"}
2021-10-13T12:01:48.208-0700    INFO    lily/chain/export       export/export.go:51     starting chain export   {"head": "[bafy2bzacecnrejmgqegtqob5s2ut2vdqimcvmaxtjngpbrjbz2niztdcdj3vq bafy2bzaceazs7kvjm5uxknrheyl3qpwrol6ioi52c6ckamm7y7hawhkijjulk]", "messages": true, "receipts": true, "stateroots": true, "final-epoch": "0"}
epochs 10000 / 10000 [==============================================>]   100.00% 4s
2021-10-13T12:01:52.259-0700    INFO    lily/chain/export       export/export.go:151    chain export complete
```

The next steps are to export the calibration network with this and implement a walk test vector.

### CODE REVIEW

File: chain/export/export.go [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+						return err
+					}
+				}
+			}
+		} else if kind == Dag {
+			if err := ce.pushNewLinks(c, todo); err != nil {
+				return err
+			}
+		} else {
+			panic("received undefined cid kind")
```
Comment (frrist) [inline chain/export/export.go:148]: this is a developer error

File: chain/export/export.go [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+
+func (ce *ChainExporter) writeContent(c cid.Cid) error {
+	blk, err := ce.store.Get(c)
+	if err != nil {
+		return err
+	}
+	return util.LdWrite(ce.out, c.Bytes(), blk.RawData())
+}
+
+type ExportOption func(ce *ChainExporter)
```
Comment (iand) [inline chain/export/export.go:173]: I think I'd rather see a single ExportOptions type with the various options as fields. That would be consistent with the rest of the codebase and a lot less code. Especially for something that is only used once.

File: chain/export/export.go [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+}
+
+func (ce *ChainExporter) Export(ctx context.Context, opts ...ExportOption) error {
+	for _, opt := range opts {
+		opt(ce)
+	}
+	log.Infow("starting chain export", "head", ce.head, "messages", ce.incMg, "receipts", ce.incRt, "stateroots", ce.incSt, "final-epoch", ce.min)
+	defer log.Info("chain export complete")
+
+	var bar *pb.ProgressBar
```
Comment (iand) [inline chain/export/export.go:54]: Do we need this? We don't have it for walks which are conceptually similar

File: commands/export.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+	Description: "Export chain from repo (requires node to be offline)",
+	Flags: []cli.Flag{
+		&cli.StringFlag{
+			Name:        "repo",
+			Usage:       "the repo to export chain from",
+			Value:       "~/.lily",
+			Destination: &chainExportFlags.repo,
+		},
+		&cli.Uint64Flag{
+			Name:        "max",
```
Comment (iand) [inline commands/export.go:42]: `to` for consistency with walk

File: commands/export.go [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+			Usage:       "inclusive max epoch to export",
+			Required:    true,
+			Destination: &chainExportFlags.max,
+		},
+		&cli.Uint64Flag{
+			Name:        "min",
```
Comment (iand) [inline commands/export.go:48]: `from` for consistency with walk

**Comment by iand** [comment]
This is synchronous. Would it be better as a job?

File: commands/export.go (hunk 4 above)
Comment (frrist) [reply commands/export.go:42]: ack will use to and from naming convention.

**Review by iand — COMMENTED** [review]
Trying to remove the request changes status

**Pull request merged**